File: python/leetcode166.py
class Solution(object):
    def fractionToDecimal(self, numerator, denominator):
        """
        :type numerator: int
        :type denominator: int
        :rtype: str
        """
        res = ''
        if numerator/denominator<0:
            res += '-'
        numerator, denominator = abs(numerator), abs(denominator)
        integer, remainder = divmod(numerator, denominator)
        res += str(integer)
        if remainder!= 0:
            res+= '.'
        i = len(res)
        table = {}
        while remainder != 0:
            if remainder not in table:
                table[remainder] = i
            else:
                i =table[remainder]
                res = res[:i]+ '('+res[i:]+')'
                return res
            remainder*= 10
            res += str(remainder//denominator)
            remainder %= denominator
            i+=1
        return res

File: leetcodec++/leetcode166.cpp
//
// Created by 黄敏辉 on 2019-08-25.
//

class Solution {
public:
    string fractionToDecimal(int numerator, int denominator) {
        if (! numerator)
            return "0";

        string res;
        if (numerator > 0 ^ denominator > 0)
            res += '-';
        long n = labs(numerator), d = labs(denominator), r = n%d;
        res += to_string(n / d);
        if (!r)
            return res;
        res += '.';
        unordered_map<long, int> rs;
        while (r){
            if (rs.find(r) != rs.end()){
                res.insert(rs[r], "(");
                res += ')';
                break;
            }
            rs[r] = res.size();
            r *= 10;
            res += to_string(r / d);
            r %= d;
        }
        return res;
    }
};